project/src/main/webapp/gmail.js

```diff
@@ -0,0 +1,365 @@
+        let emailDateValue = decodedEmail.substring(MessageIDStartIndex, MessageIDEndIndex);
+        if (emailDateValue.indexOf("X-Notifications") != -1) {
+            emailDateValue = emailDateValue.substring(0, emailDateValue.indexOf("X-Notifications"));
+        }
+        return emailDateValue;
+    }
+
+    /**
+      * Checks if the email is a valid email
+      * 
+      * Criteria 1: Isn't a Daily Insider message - breaks ML model
+      * Criteria 2: Isn't an empty email
+      * Criteria 3: Isn't a google chat message
+      */
+    isActualEmail(emailBodyValue, emailDateValue) {
+        if (emailBodyValue.indexOf("Daily Insider") != -1) {
+            return false;
+        }
+        if (emailDateValue.length <= 1) {
+            return false;
+        }
+        if (emailBodyValue.indexOf("format=flowed") != -1) {
+            return false;
+        }
+        return true;
+    }
+
+    getEmailObjects() {
+        console.log(this.emailObjects);
+    }
+
+    /**
+     * Get Message with given ID.
+     *
+     * @param  {String} messageId ID of Message to get.
+     */
+    getMessage(messageId) {
+        const getMessagePromise = new Promise((resolve) => {
+            const messageRequest = gapi.client.gmail.users.messages.get({
+                'userId': 'me',
+                'id': messageId,
+                'format': "raw"
+            });
+            messageRequest.execute(response => {
+                //Convert from base64 encoding to text.
+                const decodedEmail = atob(response.raw.replace(/-/g, '+').replace(/_/g, '/'));
+                const emailBodyValue = this.getEmailBody(decodedEmail);
+                const emailDateValue = this.getEmailDate(decodedEmail);
+                const messageID = this.getMessageID(decodedEmail);
+                if (this.isActualEmail(emailBodyValue, emailDateValue)) {
+                    this.emailObjects[messageID] = { emailDate: emailDateValue, emailBody: emailBodyValue };
+                }
+                this.getEmailObjects();
+                resolve();
+            });
+        });
+        return getMessagePromise;
+    }
+
+    //Retrieve messages using hardcoded queries and the signed-in email.
+    listMessages() {
+        this.emailObjects = {};
+        const listMessagePromise = new Promise((resolve) => {
+            const promiseArray = [];
+            var getPageOfMessages = (request, result) => {
+                request.execute(resp => {
+                    result = result.concat(resp.messages);
+                    var nextPageToken = resp.nextPageToken;
+                    if (nextPageToken) {
+                        request = gapi.client.gmail.users.messages.list({
+                            'userId': 'me',
+                            'pageToken': nextPageToken,
+                            'q': this.question
+                        });
+                        getPageOfMessages(request, result);
+                    } else {
```

Comment:
Not that I'm aware of. The promises format kinda makes it a bit complicated.

Reply:
This is a good suggestion, breaking up complicated code will make it easier to read. One way to break it up a little is to make `getPageOfMessages` a method instead of defining it inline. Just make sure that you make `promiseArray` a parameter and give it `[]` the first time you call it.